### ISSUE TITLE
about the configuration of input normalization std 

### ISSUE THREAD

**Issue by alexwq100** [open-mmlab/mmdetection#796]

**Comment by alexwq100** [comment]
hi, I notice that the configuration about the data normalization is img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True) in mmdet,
while in maskrcnn benchmark of fackbook ,the config is https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/config/defaults.py#L53
, especially the std value is 1.0. why there is a big difference here? thanks a lot

**Comment by hellock** [comment]
Image mean and std should be the same as the pretrained model. We adopt the pretrained weights from the official pytorch model zoo while maskrcnn-benchmark adopts pretrained weights from other sources. We also provides the converted weights which are the same as maskrcnn-benchmark, and the same mean and std should be adopted when using these pretrained weights.

**Comment by alexwq100** [comment]
thanks a lot again. But if these pretrained models are all trained on the imagenet dataset, I also notice that the mean difference is small , but the std difference is big, could you explain this ,thanks !

**Comment by alexwq100** [comment]
@hellock ,thanks a lot again. But if these pretrained models are all trained on the imagenet dataset, I also notice that the mean difference is small , but the std difference is big, could you explain this ,thanks !

**Comment by alexwq100** [comment]
@hellock ,thanks a lot again. But if these pretrained models are all trained on the imagenet dataset, I also notice that the mean difference is small , but the std difference is big, could you explain this ,thanks !

**Comment by hellock** [comment]
The mean and std values are decided by the pretrained models. When you are finetuning with some pretrained model, you need to follow the mean and std values used for pretraining.

**Comment by HeYuwei** [comment]
> Image mean and std should be the same as the pretrained model. We adopt the pretrained weights from the official pytorch model zoo while maskrcnn-benchmark adopts pretrained weights from other sources. We also provides the converted weights which are the same as maskrcnn-benchmark, and the same mean and std should be adopted when using these pretrained weights.


Hello,  I checked the official documentation of pytorch:

"All pre-trained models expect input images normalized in the same way, i.e. mini-batches of 3-channel RGB images of shape (3 x H x W), where H and W are expected to be at least 224. The images have to be loaded in to a range of [0, 1] and then normalized using mean = [0.485, 0.456, 0.406] and std = [0.229, 0.224, 0.225]. You can use the following transform to normalize:"

As I notice you adopt the pretrained model from torchvision, may I ask why you do not normalize the pictures to [0,1]?

**Comment by PeterVennerstrom** [comment]
MMDetection normalizes using mean and std before normalizing to [0,1].  

TorchVision normalizes to [0, 1] then normalizes using mean and std.

This makes their respective mean and std settings equivalent.

**Comment by leewi9** [comment]
thanks, how about `dict(type="Resize", img_scale=(1333, 800), keep_ratio=True),`

are 1333 and 800 also from the pretrained model?

**Comment by PeterVennerstrom** [comment]
The pretrained backbone models are typically trained on imagenet with sides between 224-448.  Most object detection models are trained with larger images, but can still benefit from pretrained weights in the backbone.  

The (1333, 800) setting is specific to the object detection model, not the backbone pretraining.

**Comment by SimonDoll** [comment]
> MMDetection normalizes using mean and std before normalizing to [0,1].
> 
> TorchVision normalizes to [0, 1] then normalizes using mean and std.
> 
> This makes their respective mean and std settings equivalent.

Thanks a lot for this clarification. 
Where is the normalization to [0,1] done? From the code I only see the mmcv.imnormalize call in the loading pipeline that subtracts the mean and divides by std afterwards.
Thanks a lot for any help!

**Comment by PeterVennerstrom** [comment]
My original statement that MMDetection normalizes using mean and std before normalizing to [0,1] was incorrect, it would have been [0,1] then mean and std like torch vision.  

In any case, the normalization is now done completely in the mmcv.imnormalize() call.  The mean and std values are scaled up by 255x compared to the torch vision values used in older MMDetection versions.    

Look in configs/_base_/datasets/coco_detection.py

New version (input [0,255]):
mean=[123.675, 116.28, 103.53]
std=[58.395, 57.12, 57.375]

Old version (input [0-1]):
mean=[0.485, 0.456, 0.406]
std=[0.229, 0.224, 0.225]

**Comment by DishantMewada** [comment]
So, does that mean If I want to change the input image sizes, I don't touch mean and std. Just change (1333, 800) to (x,y) that you want.

**Comment by GeorgePearse** [comment]
Is this normalisation now performed in MMCV? Can't see it in https://github.com/open-mmlab/mmdetection/blob/main/configs/_base_/datasets/coco_detection.py

**Comment by PeterVennerstrom** [comment]
See the [DetDataPreprocessor](https://github.com/open-mmlab/mmdetection/blob/f78af7785ada87f1ced75a2313746e4ba3149760/configs/gfl/gfl_r50_fpn_1x_coco.py#L7) setting in this GLF config.  It builds the [DetDataPreprocessor class](https://github.com/open-mmlab/mmdetection/blob/f78af7785ada87f1ced75a2313746e4ba3149760/mmdet/models/data_preprocessors/data_preprocessor.py#L31) from MMDet.  This class inherits the [MMEngine ImgDataPreprocessor](https://github.com/open-mmlab/mmengine/blob/317d8f3178a3c11188e9e9683c3dab711da75e6f/mmengine/model/base_model/data_preprocessor.py#L145) which carries out the normalization.

**Comment by GeorgePearse** [comment]
Should I set it from values based on my dataset? Or just use the COCO defaults? For context my dataset is ~45K of 1920x1080 and ~ 1 million objects. 

Or is it something that can only be determined experimentally?
